walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,117 @@
 package com.google.sps;
 
 import java.util.Collection;
+import java.util.Arrays;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.stream.Collectors;
 
 public final class FindMeetingQuery {
   public Collection<TimeRange> query(Collection<Event> events, MeetingRequest request) {
-    throw new UnsupportedOperationException("TODO: Implement this method.");
+    if(events.isEmpty() && request.getDuration() <= TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList(TimeRange.WHOLE_DAY);
+    }
+
+    if(request.getDuration() > TimeRange.WHOLE_DAY.duration()) {
+      return Arrays.asList();
+    }
+    
+    List<TimeRange> availableTimes = new ArrayList<TimeRange>();
+    List<Event> eventList = new ArrayList<>(events);
+    Event firstEvent = eventList.get(0);
+    Event lastEvent = eventList.get(eventList.size() - 1);
+
+    int eventsSkipped = 0;
+
+    if(eventList.size() == 1) {
+      if(!requestHasAttendees(request, firstEvent)) {
+        eventsSkipped++;
+      }
+    }
+
+    // Add the first time slot, from the start of the day to the start of the first event,
+    // only if there's no event that starts the day.
+    if(dayStartsWithEvent(eventList) && requestHasAttendees(request, firstEvent)) {
+      availableTimes.add(TimeRange.fromStartEnd(TimeRange.START_OF_DAY,
+        eventList.get(0).getWhen().start(), false));
+    }
+
+    for (int i = 0; i < eventList.size() - 1; i++) {
+      TimeRange currEventTimeRange = eventList.get(i).getWhen();
+      TimeRange nextEventTimeRange = eventList.get(i + 1).getWhen();
+
+      if(!requestHasAttendees(request, currEvent)) {
+        eventsSkipped ++;
+        continue;
+      }
```

Comment:
That looks strange to me. If we never modify the value of `request` and `currEvent`, this should always return the same result, right? So I'd check it outside of the loop instead. If it's false, I'd increment `eventsSkipped` by the total number of iterations you have in this loop.

Reply:
That's a mistake on my part. Instead of `currEvent` it should be `eventList.get(i)`. We check for each event if it has common attendees with our request.